src/Algebra/Graph/NonEmpty.hs

```diff
@@ -0,0 +1,563 @@
+import qualified Algebra.Graph                    as G
+import qualified Algebra.Graph.AdjacencyMap       as AM
+import qualified Algebra.Graph.Class              as C
+import qualified Algebra.Graph.HigherKinded.Class as H
+import qualified Algebra.Graph.IntAdjacencyMap    as IAM
+import qualified Algebra.Graph.Relation           as R
+import qualified Data.IntSet                      as IntSet
+import qualified Data.List.NonEmpty               as NonEmpty
+import qualified Data.Set                         as Set
+
+data NonEmptyGraph a = Vertex a
+                     | Overlay (NonEmptyGraph a) (NonEmptyGraph a)
+                     | Connect (NonEmptyGraph a) (NonEmptyGraph a)
+                     deriving (Foldable, Functor, Show, Traversable)
+
+instance NFData a => NFData (NonEmptyGraph a) where
+    rnf (Vertex  x  ) = rnf x
+    rnf (Overlay x y) = rnf x `seq` rnf y
+    rnf (Connect x y) = rnf x `seq` rnf y
+
+instance C.ToGraph (NonEmptyGraph a) where
+    type ToVertex (NonEmptyGraph a) = a
+    toGraph = foldg1 C.vertex C.overlay C.connect
+
+instance H.ToGraph NonEmptyGraph where
+    toGraph = foldg1 H.vertex H.overlay H.connect
+
+instance Ord a => Eq (NonEmptyGraph a) where
+    x == y = C.toGraph x == (C.toGraph y :: AM.AdjacencyMap a)
+
+instance Applicative NonEmptyGraph where
+    pure  = Vertex
+    (<*>) = ap
+
+instance Monad NonEmptyGraph where
+    return  = pure
+    g >>= f = foldg1 f Overlay Connect g
+
+-- | Construct the graph comprising /a single isolated vertex/. An alias for the
+-- constructor 'Vertex'.
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- 'hasVertex' x (vertex x) == True
+-- 'hasVertex' 1 (vertex 2) == False
+-- 'vertexCount' (vertex x) == 1
+-- 'edgeCount'   (vertex x) == 0
+-- 'size'        (vertex x) == 1
+-- @
+vertex :: a -> NonEmptyGraph a
+vertex = Vertex
+
+-- | Construct the graph comprising /a single edge/.
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- edge x y               == 'connect' ('vertex' x) ('vertex' y)
+-- 'hasEdge' x y (edge x y) == True
+-- 'edgeCount'   (edge x y) == 1
+-- 'vertexCount' (edge 1 1) == 1
+-- 'vertexCount' (edge 1 2) == 2
+-- @
+edge :: a -> a -> NonEmptyGraph a
+edge u v = connect (vertex u) (vertex v)
+
+-- | /Overlay/ two graphs. An alias for the constructor 'Overlay'. This is an
+-- idempotent, commutative and associative operation with the identity 'empty'.
+-- Complexity: /O(1)/ time and memory, /O(s1 + s2)/ size.
+--
+-- @
+-- 'hasVertex' z (overlay x y) == 'hasVertex' z x || 'hasVertex' z y
+-- 'vertexCount' (overlay x y) >= 'vertexCount' x
+-- 'vertexCount' (overlay x y) <= 'vertexCount' x + 'vertexCount' y
+-- 'edgeCount'   (overlay x y) >= 'edgeCount' x
+-- 'edgeCount'   (overlay x y) <= 'edgeCount' x   + 'edgeCount' y
+-- 'size'        (overlay x y) == 'size' x        + 'size' y
+-- 'vertexCount' (overlay 1 2) == 2
+-- 'edgeCount'   (overlay 1 2) == 0
+-- @
+overlay :: NonEmptyGraph a -> NonEmptyGraph a -> NonEmptyGraph a
```

Comment:
`Data.List.NonEmpty` also misses appends like:

```haskell
appendl :: [a] -> NonEmpty a -> NonEmpty a
appendr :: NonEmpty a -> [a] -> NonEmpty a
```

good to keep in mind, but coming up with good names is difficult.

Reply:
Good point. For `overlay` we could probably provide only one `overlay1` function due to commutativity, but for `connect` we'll need both.

I decided not to add these functions for now, because I don't expect users to mix non-empty and general graphs in a single project. We'll see if there will be any demand.

In general, this seems to be a perfect use-case for subtyping. Just for the sake of curiousity I'll give it a try in my draft Scala implementation.